### DISCUSSION TITLE
为什么 VoidField 没有 `errors` 等 feedback 字段？

### DISCUSSION THREAD

**Discussion by JustDs** [alibaba/formily#1685] [Ideas]
个人认为 void 类型字段因为没有值，没有 `value` 和 `validator` 之类的字段是可以理解的，但是 `feedbacks`、`errors`、`warnings`、`successes` 这些字段是不是可以考虑保留？

比如对某一块内容而非某个单一字段显示报错信息的场合下，会比较有用，现在会存在找不到地方存放这些状态的问题。

**Comment by janryWang** [comment]
反模式，一个虚数据字段，怎么可能会存在校验，讲都讲不清楚

**Reply by JustDs** [reply]
行吧，那就还是给实际例子：

https://codesandbox.io/s/vibrant-tu-j5j5b?file=/App.tsx

需要在提交时校验左边框内值小于右边框内值，否则在 range 字段下给出报错信息，应该如何处理？

**Reply by janryWang** [reply]
为啥不直接用Range组件？还有，错误消息，直接显示在min和max也没问题呀，就是个联动校验而已

**Reply by JustDs** [reply]
且不说 UI 展示需要是这样，数字的 Range 组件你这哪有？

错误消息直接显示在 min 和 max 下会被折断，UI 要求显示在整个字段下而非单个框下。

**Reply by janryWang** [reply]
antd本身就有Range组件呀，双滑条 Slider

**Comment by janryWang** [comment]
两个方案：
1. 使用Antd Slider双滑条
2. 自定义组件，类似于DatePicker.Range，然后name用解构
3. 手动设置x-decorator-props.feedbackText/feedbackStatus

**Reply by JustDs** [reply]
设计稿就是两个输入框，总不能因为我用了 formily 就得让人家改稿子吧？

用 object 解构的，能不能来个例子？就在上面 sandbox 改就好，不用加我说的额外校验的部分，表单能正常工作就行。

**Reply by janryWang** [reply]
感觉你好难……一边被交互打压，一边被后端打压

**Reply by JustDs** [reply]
但是问题总归要解决的嘛，针对某一块区域报错这点就交互逻辑上来讲至少不是反范式的，我觉得确实是应当解决的问题啊。

**Reply by janryWang** [reply]
刚刚看了下，并不支持嵌套解构关系，这种太复杂了，我推荐你实现一个类似DatePicker.RangePicker的组件，value/onChange 传递数据改成[min,max]就行了，然后再参考 https://v2.formilyjs.org/zh-CN/guide/advanced/destructor

**Reply by janryWang** [reply]
总之，不推荐void区块处理错误消息的东西，讲不清楚它的定位

**Reply by JustDs** [reply]
我的实际业务场景比这个还要再复杂点，一个 RangePicker 之类的的组件可能还有点难完美满足要求。

其实问题就在于「针对某块东西」报错这件事本身是很合理的，如果这块东西是 void 的，现在 formily 就搞不定了，非得用解构之类的方式强行把它变成不 void 的。结构简单结构还行，那如果是对某个卡片、某个区域的报错呢？

**Reply by JustDs** [reply]
Void 区块其实就是「一块东西」啊，针对「这块东西」需要有报错提示，怎么就解释不清定位了？

**Reply by JustDs** [reply]
另外我说的是「错误提示」而不是「校验」。校验会产生错误提示，但是错误提示并不一定来自校验啊，也可以是手动设置进去的。

**Reply by janryWang** [reply]
这只是你有这个需求吧，合理的就是报错都是在数据字段展示报错，而不是在什么区块里报错，这是最佳实践！还有你说的什么展示不下的问题，你把NumberPicker弄长一点不就行了

**Reply by JustDs** [reply]
这事情就是联动校验啊，几个字段综合起来判断发现有问题要报错，你说该挂在哪个字段上？本来 `x-validator` 不让去取别的字段我就已经很想吐槽了，现在手动设置也没地方放。

展示不下是因为本身 range 的错误挂在任何单独一边就是不合理的，这是弄长不弄长的问题么。

**Reply by janryWang** [reply]
> 这事情就是联动校验啊，几个字段综合起来判断发现有问题要报错，你说该挂在哪个字段上？本来 `x-validator` 不让去取别的字段我就已经很想吐槽了，现在手动设置也没地方放。
> 
> 展示不下是因为本身 range 的错误挂在任何单独一边就是不合理的，这是弄长不弄长的问题么。

那你为啥不用object，说实话，用object，对后端来说更适合领域建模，说真的，该说的方法都说了，你吐槽归吐槽，不合理的需求，我是不接受的，除非有很多人遇到和你一样的场景，就目前而言，我还是第一次遇到

**Comment by janryWang** [comment]
你说你找不到地方存错误提示，直接x-decorator-props就能存 https://codesandbox.io/s/pensive-austin-grjns?file=/App.tsx

**Comment by liuweiGL** [comment]
这么简单的需求搞的这么复杂干嘛：https://codesandbox.io/s/angry-mclaren-2490o?file=/App.tsx

**Reply by janryWang** [reply]
对啊，这个方案我也说了，他就是不喜欢，我也是醉了
